### DISCUSSION TITLE
Add missing & update existing validation attributes

### DISCUSSION THREAD

**Discussion by Klaas058** [spatie/laravel-data#570] [Ideas]
I cross-referenced the [validation attributes](https://github.com/spatie/laravel-data/blob/main/docs/advanced-usage/validation-attributes.md) with the [Laravel 9.x available validation rules](https://laravel.com/docs/9.x/validation#available-validation-rules) and found that there are 18 validation rules that are not implemented as validation attributes yet.

I couldn't find information about which Laravel version this package needs to support, so depending on the version support, I would like to add the following:

- [x] declined_if [8.71.0](https://github.com/laravel/framework/releases/tag/v8.71.0)
- [x] declined [8.71.0](https://github.com/laravel/framework/releases/tag/v8.71.0)
- [x] mac [8.77.0](https://github.com/laravel/framework/releases/tag/v8.77.0)
- [x] required_array_keys [8.82.0](https://github.com/laravel/framework/releases/tag/v8.82.0)
- [x] exclude_with [9.6.0](https://github.com/laravel/framework/releases/tag/v9.6.0)
- [x] doesnt_start_with [9.17.0](https://github.com/laravel/framework/releases/tag/v9.17.0)
- [x] doesnt_end_with [9.24.0](https://github.com/laravel/framework/releases/tag/v9.24.0)
- [x] min_digits [9.26.0](https://github.com/laravel/framework/releases/tag/v9.26.0)
- [x] max_digits [9.26.0](https://github.com/laravel/framework/releases/tag/v9.26.0)
- [x] uppercase [9.40.0](https://github.com/laravel/framework/releases/tag/v9.40.0)
- [x] lowercase [9.40.0](https://github.com/laravel/framework/releases/tag/v9.40.0)
- [x] ascii [9.44.0](https://github.com/laravel/framework/releases/tag/v9.44.0)
- [x] decimal [9.45.0](https://github.com/laravel/framework/releases/tag/v9.45.0)
- [ ] missing_with_all [9.49.0](https://github.com/laravel/framework/releases/tag/v9.49.0)
- [ ] missing_with [9.49.0](https://github.com/laravel/framework/releases/tag/v9.49.0)
- [ ] missing_unless [9.49.0](https://github.com/laravel/framework/releases/tag/v9.49.0)
- [ ] missing_if [9.49.0](https://github.com/laravel/framework/releases/tag/v9.49.0)
- [ ] missing [9.49.0](https://github.com/laravel/framework/releases/tag/v9.49.0)

I would like to open PR's to add these validation rules.

Also I found that [exclude](https://github.com/spatie/laravel-data/blob/main/src/Attributes/Validation/Exclude.php) is implemented but not referenced in the docs, the sibling attributes are referenced. The implementation of the [exclude](https://github.com/spatie/laravel-data/blob/main/src/Attributes/Validation/Exclude.php) however needs an instance of the `ExcludeIf` class to be passed as a parameter. It would make more sense to implement it like the [RequiredIf](https://github.com/spatie/laravel-data/blob/main/src/Attributes/Validation/RequiredIf.php) is implemented

- [ ] Update the implementation of `ExcludeIf`

Lastly, the `alpha`, `alpha_dash`, and `alpha_numeric` received an update in [9.49.0](https://github.com/laravel/framework/releases/tag/v9.49.0) to add the `ascii` parameter. These could also be updated.

- [ ] Update the implementations of `Alpha`, `AlphaDash`, `AlphaNumeric`

There are probably more to update, but these are the ones I found.

**Comment by rubenvanassche** [comment]
Looks like a great idea! Already saw some PR's coming in, thanks!
